In [1]:
from utils import max_features, maxlen, embed_size_fastText, embed_size_glove, embed_size_glove_twitter
from utils import batch_size, epochs, run_5fold, write_prediction
from model import get_textCNN
import pickle
import tensorflow as tf
import pandas as pd

/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ec2-user/.config/matplotlib/matplotlibrc", line #2
  (fname, cnt))
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ec2-user/.config/matplotlib/matplotlibrc", line #3
  (fname, cnt))
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into whic

In [2]:
fastText = pickle.load(open("input/fastText.300d.pickle", "rb"))
glove = pickle.load(open("input/glove.300d.pickle", "rb"))
glove_twitter = pickle.load(open("input/glove.twitter.200d.pickle", "rb"))

In [3]:
x_train = pickle.load(open("input/x_train.pickle", "rb"))
y_train = pickle.load(open("input/y_train.pickle", "rb"))
x_test = pickle.load(open("input/x_test.pickle", "rb"))
submission = pd.read_csv("input/sample_submission.csv")

## check different embedding

In [8]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "num_filter": 50,
    "reg": 0.0,
    "dropout": 0.2
}

In [5]:
embeddings = [(fastText, embed_size_fastText, "fastText"), (glove, embed_size_glove, "glove"), (glove_twitter, embed_size_glove_twitter, "glove_twitter")]

In [6]:
predictions = {}

In [9]:
for embed in embeddings:
    print ("\n********************{}********************".format(embed[2]))
    params["embedding_matrix"] = embed[0]
    params["embed_size"] = embed[1]
    y_preds = run_5fold(x_train, y_train, x_test, "loss", get_textCNN, **params)
    predictions[embed[2]] = y_preds


********************fastText********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 68s - loss: 0.0749 - acc: 0.9771 - auc_roc: 0.9159 - val_loss: 0.0476 - val_acc: 0.9822 - val_auc_roc: 0.9626
Epoch 2/50
 - 67s - loss: 0.0416 - acc: 0.9843 - auc_roc: 0.9713 - val_loss: 0.0469 - val_acc: 0.9821 - val_auc_roc: 0.9760
Epoch 3/50
 - 67s - loss: 0.0337 - acc: 0.9870 - auc_roc: 0.9794 - val_loss: 0.0488 - val_acc: 0.9816 - val_auc_roc: 0.9816
Epoch 00003: early stopping
val loss: 0.04879099277389305
val acc: 0.9816178171253309
val auc roc: 0.9816154052504137

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 68s - loss: 0.0781 - acc: 0.9771 - auc_roc: 0.9151 - val_loss: 0.0474 - val_acc: 0.9826 - val_auc_roc: 0.9622
Epoch 2/50
 - 67s - loss: 0.0413 - acc: 0.9845 - auc_roc: 0.9711 - val_loss: 0.0457 - val_acc: 0.9828 - val_auc_roc: 0.9761
Epoch 3/50
 - 67s - loss: 0.0332 - acc: 0.9873 - auc_roc: 0.9794 - val_loss: 0.0478 - val_acc: 0.9825 - val_a

Epoch 2/50
 - 53s - loss: 0.0458 - acc: 0.9831 - auc_roc: 0.9585 - val_loss: 0.0471 - val_acc: 0.9824 - val_auc_roc: 0.9662
Epoch 3/50
 - 53s - loss: 0.0391 - acc: 0.9851 - auc_roc: 0.9711 - val_loss: 0.0476 - val_acc: 0.9821 - val_auc_roc: 0.9744
Epoch 00003: early stopping
val loss: 0.047609251771235914
val acc: 0.9820872612029992
val auc roc: 0.9743622337926888



In [11]:
from utils import eval_results

In [12]:
eval_results(predictions)

key: fastText
loss: 0.04788035589682653 (+/- 0.0005406496585757341)
acc: 0.9825448492123818 (+/- 0.0005274444695278663)
roc_auc: 0.9814734088264092 (+/- 0.0015244724636944255)



key: glove
loss: 0.04706004063530118 (+/- 0.0006838306930178759)
acc: 0.9824341330567711 (+/- 0.0003545645226532429)
roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)



key: glove_twitter
loss: 0.048162465279300334 (+/- 0.0003795115630879772)
acc: 0.9823714628808649 (+/- 0.0003519317806489231)
roc_auc: 0.9769190395744314 (+/- 0.0024952235885009597)





##### result: glove
- fastText: 

loss: 0.04788035589682653 (+/- 0.0005406496585757341)

acc: 0.9825448492123818 (+/- 0.0005274444695278663)

roc_auc: 0.9814734088264092 (+/- 0.0015244724636944255)

- glove: 

loss: 0.04706004063530118 (+/- 0.0006838306930178759)

acc: 0.9824341330567711 (+/- 0.0003545645226532429)

roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)

- glove.twitter: 

loss: 0.048162465279300334 (+/- 0.0003795115630879772)

acc: 0.9823714628808649 (+/- 0.0003519317806489231)

roc_auc: 0.9769190395744314 (+/- 0.0024952235885009597)


In [17]:
opt = predictions['glove']['pred']

## check num_filter
(suggest values: 100~600: test 100, 300, 500)

In [15]:
filters = [100, 300, 500]  # for the result of 300, refer to the previous log

In [13]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "reg": 0.0,
    "dropout": 0.2,
    "embedding_matrix": glove,
    "embed_size": embed_size_glove
}

In [19]:
for f in filters:
    print ("\n********************f = {}********************".format(f))
    params["num_filter"] = f
    y_preds = run_5fold(x_train, y_train, x_test, "loss", get_textCNN, **params)
    predictions[f] = y_preds


********************f = 100********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 89s - loss: 0.0783 - acc: 0.9775 - auc_roc: 0.9084 - val_loss: 0.0537 - val_acc: 0.9805 - val_auc_roc: 0.9558
Epoch 2/50
 - 88s - loss: 0.0426 - acc: 0.9842 - auc_roc: 0.9660 - val_loss: 0.0502 - val_acc: 0.9816 - val_auc_roc: 0.9717
Epoch 3/50
 - 88s - loss: 0.0355 - acc: 0.9866 - auc_roc: 0.9756 - val_loss: 0.0485 - val_acc: 0.9822 - val_auc_roc: 0.9782
Epoch 4/50
 - 87s - loss: 0.0287 - acc: 0.9891 - auc_roc: 0.9804 - val_loss: 0.0511 - val_acc: 0.9823 - val_auc_roc: 0.9818
Epoch 00004: early stopping
val loss: 0.051138400882552566
val acc: 0.9823123709266133
val auc roc: 0.9818147175106232

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 88s - loss: 0.0851 - acc: 0.9770 - auc_roc: 0.9041 - val_loss: 0.0481 - val_acc: 0.9825 - val_auc_roc: 0.9532
Epoch 2/50
 - 88s - loss: 0.0430 - acc: 0.9841 - auc_roc: 0.9644 - val_loss: 0.0477 - val_acc: 0.9828 - val_a

 - 261s - loss: 0.1130 - acc: 0.9775 - auc_roc: 0.8779 - val_loss: 0.0599 - val_acc: 0.9823 - val_auc_roc: 0.9275
Epoch 2/50
 - 260s - loss: 0.0512 - acc: 0.9841 - auc_roc: 0.9431 - val_loss: 0.0601 - val_acc: 0.9811 - val_auc_roc: 0.9522
Epoch 00002: early stopping
val loss: 0.060051398297976434
val acc: 0.9811106752936923
val auc roc: 0.9521632153691897



In [20]:
eval_results(predictions)

key: fastText
loss: 0.04788035589682653 (+/- 0.0005406496585757341)
acc: 0.9825448492123818 (+/- 0.0005274444695278663)
roc_auc: 0.9814734088264092 (+/- 0.0015244724636944255)



key: glove
loss: 0.04706004063530118 (+/- 0.0006838306930178759)
acc: 0.9824341330567711 (+/- 0.0003545645226532429)
roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)



key: glove_twitter
loss: 0.048162465279300334 (+/- 0.0003795115630879772)
acc: 0.9823714628808649 (+/- 0.0003519317806489231)
roc_auc: 0.9769190395744314 (+/- 0.0024952235885009597)



key: 100
loss: 0.049184090981469 (+/- 0.002029732246391881)
acc: 0.9822503025945395 (+/- 0.0006452397105241698)
roc_auc: 0.9793636124334473 (+/- 0.0015355110356199272)



key: 300
loss: 0.05324313215723492 (+/- 0.003979328743352576)
acc: 0.9819599517713336 (+/- 0.0007649242349705409)
roc_auc: 0.9709089147040133 (+/- 0.004742504171523545)



key: 500
loss: 0.05876650591737374 (+/- 0.004129826861398923)
acc: 0.9815139566836362 (+/- 0.0008138840972856712)
roc

###### result: 

- f = 50

loss: 0.04706004063530118 (+/- 0.0006838306930178759)

acc: 0.9824341330567711 (+/- 0.0003545645226532429)

roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)


- f = 100

loss: 0.049184090981469 (+/- 0.002029732246391881)

acc: 0.9822503025945395 (+/- 0.0006452397105241698)

roc_auc: 0.9793636124334473 (+/- 0.0015355110356199272)

- f = 300

loss: 0.05324313215723492 (+/- 0.003979328743352576)

acc: 0.9819599517713336 (+/- 0.0007649242349705409)

roc_auc: 0.9709089147040133 (+/- 0.004742504171523545)

- f = 500

loss: 0.05876650591737374 (+/- 0.004129826861398923)

acc: 0.9815139566836362 (+/- 0.0008138840972856712)

roc_auc: 0.9616913396704657 (+/- 0.00578077091317378)

In [21]:
opt = opt

## check kernel_size
(suggest value: [3, 5, 7, 9])

In [22]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "dropout": 0.2,
    "num_filter": 50,
    "reg": 0.0,
    "embedding_matrix": glove,
    "embed_size": embed_size_glove
}

In [23]:
kernels = [5, 7, 9]  # for 3, see result above

In [24]:
for k in kernels:
    print ("\n********************k = {}********************".format(k))
    params["kernel_size"] = k
    y_preds = run_5fold(x_train, y_train, x_test, "loss", get_textCNN, **params)
    predictions[k] = y_preds


********************k = 5********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 68s - loss: 0.0833 - acc: 0.9757 - auc_roc: 0.8980 - val_loss: 0.0477 - val_acc: 0.9824 - val_auc_roc: 0.9555
Epoch 2/50
 - 67s - loss: 0.0433 - acc: 0.9838 - auc_roc: 0.9663 - val_loss: 0.0465 - val_acc: 0.9824 - val_auc_roc: 0.9721
Epoch 3/50
 - 67s - loss: 0.0364 - acc: 0.9861 - auc_roc: 0.9759 - val_loss: 0.0463 - val_acc: 0.9829 - val_auc_roc: 0.9785
Epoch 4/50
 - 67s - loss: 0.0304 - acc: 0.9883 - auc_roc: 0.9806 - val_loss: 0.0514 - val_acc: 0.9817 - val_auc_roc: 0.9821
Epoch 00004: early stopping
val loss: 0.051383588081492765
val acc: 0.9817431485299946
val auc roc: 0.9821235690093724

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 68s - loss: 0.0642 - acc: 0.9789 - auc_roc: 0.9272 - val_loss: 0.0478 - val_acc: 0.9821 - val_auc_roc: 0.9668
Epoch 2/50
 - 67s - loss: 0.0430 - acc: 0.9838 - auc_roc: 0.9739 - val_loss: 0.0467 - val_acc: 0.9824 - val_auc

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 68s - loss: 0.0695 - acc: 0.9776 - auc_roc: 0.9166 - val_loss: 0.0469 - val_acc: 0.9826 - val_auc_roc: 0.9620
Epoch 2/50
 - 67s - loss: 0.0431 - acc: 0.9839 - auc_roc: 0.9708 - val_loss: 0.0464 - val_acc: 0.9827 - val_auc_roc: 0.9755
Epoch 3/50
 - 67s - loss: 0.0362 - acc: 0.9861 - auc_roc: 0.9787 - val_loss: 0.0484 - val_acc: 0.9828 - val_auc_roc: 0.9806
Epoch 00003: early stopping
val loss: 0.04836859876644703
val acc: 0.9827504972545861
val auc roc: 0.9805701681005781



In [25]:
eval_results(predictions)

key: fastText
loss: 0.04788035589682653 (+/- 0.0005406496585757341)
acc: 0.9825448492123818 (+/- 0.0005274444695278663)
roc_auc: 0.9814734088264092 (+/- 0.0015244724636944255)



key: glove
loss: 0.04706004063530118 (+/- 0.0006838306930178759)
acc: 0.9824341330567711 (+/- 0.0003545645226532429)
roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)



key: glove_twitter
loss: 0.048162465279300334 (+/- 0.0003795115630879772)
acc: 0.9823714628808649 (+/- 0.0003519317806489231)
roc_auc: 0.9769190395744314 (+/- 0.0024952235885009597)



key: 100
loss: 0.049184090981469 (+/- 0.002029732246391881)
acc: 0.9822503025945395 (+/- 0.0006452397105241698)
roc_auc: 0.9793636124334473 (+/- 0.0015355110356199272)



key: 300
loss: 0.05324313215723492 (+/- 0.003979328743352576)
acc: 0.9819599517713336 (+/- 0.0007649242349705409)
roc_auc: 0.9709089147040133 (+/- 0.004742504171523545)



key: 500
loss: 0.05876650591737374 (+/- 0.004129826861398923)
acc: 0.9815139566836362 (+/- 0.0008138840972856712)
roc

##### result

- k = 3

loss: 0.04706004063530118 (+/- 0.0006838306930178759)

acc: 0.9824341330567711 (+/- 0.0003545645226532429)

roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)


- k = 5

loss: 0.04892370349115508 (+/- 0.0015085776549605152)

acc: 0.9822179272184026 (+/- 0.0003295982559492814)

roc_auc: 0.9821142691917852 (+/- 0.00081992394342549)

- k = 7

loss: 0.04749907681010568 (+/- 0.0012294479215664285)

acc: 0.9824278642307375 (+/- 0.0005478350384179172)

roc_auc: 0.9783617831202976 (+/- 0.002081806086928983)

- k = 9

loss: 0.04815169093899214 (+/- 0.0018300053199111733)

acc: 0.9826973381712515 (+/- 0.00040480313870952647)

roc_auc: 0.9806889840881015 (+/- 0.0017318996390097085)


In [26]:
opt = opt

## check drop out
(suggest value 0.1~0.5: test 0.1, 0.2, 0.5)

In [28]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "kernel_size": 3,
    "num_filter": 50,
    "reg": 0.0,
    "embedding_matrix": glove,
    "embed_size": embed_size_glove
}

In [29]:
dropouts = [0.1, 0.5] # 0.2 see previous result

In [30]:
for d in dropouts:
    print ("\n********************d = {}********************".format(d))
    params["dropout"] = d
    y_preds = run_5fold(x_train, y_train, x_test, "loss", get_textCNN, **params)
    predictions[d] = y_preds


********************d = 0.1********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 68s - loss: 0.0830 - acc: 0.9757 - auc_roc: 0.8956 - val_loss: 0.0481 - val_acc: 0.9821 - val_auc_roc: 0.9555
Epoch 2/50
 - 67s - loss: 0.0410 - acc: 0.9845 - auc_roc: 0.9669 - val_loss: 0.0468 - val_acc: 0.9828 - val_auc_roc: 0.9729
Epoch 3/50
 - 67s - loss: 0.0327 - acc: 0.9875 - auc_roc: 0.9769 - val_loss: 0.0486 - val_acc: 0.9822 - val_auc_roc: 0.9795
Epoch 00003: early stopping
val loss: 0.048577008253748694
val acc: 0.9821922573422481
val auc roc: 0.9795493684348391

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 68s - loss: 0.0641 - acc: 0.9791 - auc_roc: 0.9293 - val_loss: 0.0474 - val_acc: 0.9823 - val_auc_roc: 0.9674
Epoch 2/50
 - 67s - loss: 0.0399 - acc: 0.9849 - auc_roc: 0.9749 - val_loss: 0.0464 - val_acc: 0.9831 - val_auc_roc: 0.9787
Epoch 3/50
 - 67s - loss: 0.0318 - acc: 0.9879 - auc_roc: 0.9812 - val_loss: 0.0487 - val_acc: 0.9826 - val_a

In [31]:
eval_results(predictions)

key: fastText
loss: 0.04788035589682653 (+/- 0.0005406496585757341)
acc: 0.9825448492123818 (+/- 0.0005274444695278663)
roc_auc: 0.9814734088264092 (+/- 0.0015244724636944255)



key: glove
loss: 0.04706004063530118 (+/- 0.0006838306930178759)
acc: 0.9824341330567711 (+/- 0.0003545645226532429)
roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)



key: glove_twitter
loss: 0.048162465279300334 (+/- 0.0003795115630879772)
acc: 0.9823714628808649 (+/- 0.0003519317806489231)
roc_auc: 0.9769190395744314 (+/- 0.0024952235885009597)



key: 100
loss: 0.049184090981469 (+/- 0.002029732246391881)
acc: 0.9822503025945395 (+/- 0.0006452397105241698)
roc_auc: 0.9793636124334473 (+/- 0.0015355110356199272)



key: 300
loss: 0.05324313215723492 (+/- 0.003979328743352576)
acc: 0.9819599517713336 (+/- 0.0007649242349705409)
roc_auc: 0.9709089147040133 (+/- 0.004742504171523545)



key: 500
loss: 0.05876650591737374 (+/- 0.004129826861398923)
acc: 0.9815139566836362 (+/- 0.0008138840972856712)
roc

##### result

- d = 0.1

loss: 0.0483672331288506 (+/- 0.0003929476902811908)

acc: 0.9824289087613398 (+/- 0.00014670238178248406)

roc_auc: 0.9803635218457915 (+/- 0.0019494778803073492)

- d = 0.2

loss: 0.04706004063530118 (+/- 0.0006838306930178759)

acc: 0.9824341330567711 (+/- 0.0003545645226532429)

roc_auc: 0.9810236005723141 (+/- 0.0018057481516012563)

- d = 0.5

loss: 0.04736179204627119 (+/- 0.001138485383340357)

acc: 0.9824226486718255 (+/- 0.0006514104654866455)

roc_auc: 0.9806074253556515 (+/- 0.0035496231203654776)


In [32]:
opt = opt

#### tuned params

In [14]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "kernel_size": 3,
    "num_filter": 50,
    "reg": 0.0,
    "embedding_matrix": glove,
    "embed_size": embed_size_glove,
    "dropout": 0.2
}

### write result

In [34]:
write_prediction(opt, submission, "output/textCNN.csv")